yax/shared/utilities/taxidtool.py

```diff
@@ -0,0 +1,428 @@
+
+
+def parse_nodes_dmp(nodes_fp):
+    results = {}
+    with open(nodes_fp, mode='r') as fh:
+        lines = fh.readlines()
+
+    for line in lines:
+        taxid, parent_taxid, rank, _ = \
+            line.split("\t|\t", 3)
+        results[taxid] = NodeRecord(taxid, parent_taxid, rank)
+
+    return results
+
+
+TaxIDDataRecord = collections.namedtuple(
+    "TaxIDDataRecord", (
+        "taxid",
+        "sci_name",
+        "assoc_gis",
+        "children",
+        "parents",
+        "rank",
+        "num_gis_assoc"))
+
+
+def parse_taxid_data(input_taxid_data_fp):
+    """Reads in provided taxid_data file to python dictionary.
+
+    Reads precomputed taxid_data file back into a python dictionary. This new
+    dictionary will represent whatever tree was used to build the taxid_data
+    file.
+
+    Parameters
+    ----------
+    input_taxid_data_fp : str
+        location of precomputed taxid_data file
+
+    Returns
+    -------
+    taxid_data : dict
+        contains all taxids found in provided taxid_data
+
+    See Also
+    --------
+    write_taxid_data
+
+    """
+    num_gis_assoc = 0
+    results = {}
+    with open(input_taxid_data_fp, mode='r') as fh:
+        lines = fh.readlines()
+
+    for line in lines:
+        taxid, sci_name, assoc_gis, children, ancestors, rank = \
+            line.rstrip().split("\t")
+
+        assoc_gis = json.loads(assoc_gis)
+        children = json.loads(children)
+        ancestors = json.loads(ancestors)
+
+        results[taxid] = TaxIDDataRecord(taxid, sci_name, assoc_gis,
+                                         children, ancestors, rank,
+                                         num_gis_assoc)
+
+    return results
+
+
+def build_taxid_data(nodes_fp, names_fp, gi_taxid_nucl_fp):
+    """Builds taxid_data.
+
+    Based on NCBI taxa.dmp files the taxid_data
+
+    Parameters
+    ----------
+    nodes_fp : str
+        location of nodes.dmp file
+    names_fp : str
+        location of names.dmp file
+    gi_taxid_nucl_fp : str
+        location of gi_taxid_nucl.dmp file
+
+    Returns
+    -------
+    dict
+        Big Dictionary
+
+    See Also
+    --------
+    parse_nodes_dmp
+    parse_names_dmp
+    parse_gi_taxid_dmp
+
+    """
+    nodes = parse_nodes_dmp(nodes_fp)
+    names = parse_names_dmp(names_fp)
+    taxid_gi = parse_gi_taxid_dmp(gi_taxid_nucl_fp)
+
+    if not (len(nodes) == len(names) >= len(taxid_gi)):
+        raise ValueError("NCBI dump files do not make sense")
+
+    taxid_data = {}
+    for taxid in nodes:
+        children = []
+        parents = []
+        taxid_data[taxid] = TaxIDDataRecord(taxid,
+                                            names[taxid],
+                                            taxid_gi.get(taxid, []),
+                                            children,
+                                            parents,
+                                            nodes[taxid].rank,
+                                            0)
```

Comment:
You won't be able to update `num_gis_assoc` after the fact, so this will always be zero. If we need to delay instantiation that is fine, but wouldn't a `len(record.assoc_gis)` be just as clear?